EventedFileUpdateChecker boots once per process

schneems opened the pull request:
We need one file checker booted per process as talked about in #24990. Before we do a check to see if any updates have been registered by the listener we first check to make sure that the current process has booted a listener.

We are intentionally not starting a listener when the checker is created. This way we can avoid #25259 in which puma warns of multiple threads created before fork. As written the listener for each process will be invoked by the `ActionDispatch::Executor` middleware when the `updated?` method is called. This is the first middleware on the stack and will be invoked before application code is read into memory.

The downside of this approach is that the API is a little less obvious. I.e. that you have to call `updated?` to get the listener to start is not intuitive. We could make `boot!` not private if we want to make the API a little nicer. Alternatively we could boot when the checker is initialized however this reintroduces the puma threads warning, and also means that in cases of `rails server` or when using `preload!` that we have extra threads notifying of changes on a process that we don't care about.

[close #24990] [close #25259]

schneems committed:
EventedFileUpdateChecker boots once per process

We need one file checker booted per process as talked about in #24990. Before we do a check to see if any updates have been registered by the listener we first check to make sure that the current process has booted a listener.

We are intentionally not starting a listener when the checker is created. This way we can avoid #25259 in which puma warns of multiple threads created before fork. As written the listener for each process will be invoked by the `ActionDispatch::Executor` middleware when the `updated?` method is called. This is the first middleware on the stack and will be invoked before application code is read into memory.

The downside of this approach is that the API is a little less obvious. I.e. that you have to call `updated?` to get the listener to start is not intuitive. We could make `boot!` not private if we want to make the API a little nicer. Alternatively we could boot when the checker is initialized however this reintroduces the puma threads warning, and also means that in cases of `rails server` or when using `preload!` that we have extra threads notifying of changes on a process that we don't care about.

[close #24990] [close #25259]

schneems committed:
[ci skip] document EventedFileUpdateChecker

schneems commented:
r? @matthewd

schneems committed:
Listen earlier in EventedFileUpdateChecker

Some files like routes.rb may be very large and vary between the initialization of the app and the first request. In these scenarios if we are using a forked process we cannot rely on the files to be unchanged between when the code is booted and the listener is started. 

For that reason we start a listener on the main process immediately, when we detect that a process does not have a listener started we force the updated state to be true, so we are guaranteed to catch any changes made between the code initialization and the fork.

schneems commented:
Updated with logic to ensure we're never in a situation where a process has stale data.

sgrif commented on activesupport/lib/active_support/evented_file_update_checker.rb:
```diff
@@ -28,11 +57,12 @@ def initialize(files, dirs = {}, &block)
     end
 
     def updated?
+      boot! unless @pid_hash[Process.pid]
```
Does this need to be atomic? e.g.

```
pid = Process.pid
@pid_hash.compute_if_absent(pid) { boot! }
@pid_hash[pid] = true
```

schneems replied:
In practice it's fine if we accidentally boot more than one listener per process, we want to make sure we boot at least once though. `Concurrent::Hash` doesn't have a `compute_if_absent` locking method

thedarkone replied:
> `Concurrent::Hash` doesn't have a `compute_if_absent` locking method

The method is from `Concurrent::Map` (`Conc::Map` is a `Hash`-like data structure with a few concurrency methods thrown in, `Concurrecy::Hash` is just a synchronized carbon copy of Ruby `Hash`).

Do we even need hash for this? Can't we just have `@listen_started_on_pid` and check `Process.pid == @listen_started_on_pid`, with `@listen_started_on_pid = Process.pid` in `boot!`?

sgrif replied:
We should still use an atomic even if it's not a hash. While it's technically fine to have more than one listener, it's unnecessary and easy to avoid.

sgrif commented:
I definitely think we need the API to be more explicit here. The fact that `updated?` is what causes the checker to boot makes the code path that causes this to work to become really nonlocal. (I was actually convinced that this didn't fix the issue at first, took about 20 minutes to see why it does).

It'd be good to add an integration test for this to the railties test suite as well.

schneems commented:
We need some hook that we know will be called in each process, in this case I know `updated?` will be executed in the middleware. If we had some kind of `on_fork` callback this logic would all be much simpler, but nothing like that exists and I don't feel like monkey patching `fork` is a great idea.

> It'd be good to add an integration test for this to the railties test suite as well.

I can look into that.

schneems committed:
Test how evented file checker handles forks

Pretty proud of this. We are testing distributed processes synchronized via pipes which makes it deterministic. Pretty cool.

We boot a listener in the parent process we then fork. Before we touch the file we verify the fork is booted using pipes. Then the parent process will touch the file while the fork waits on a pipe. Once the parent process signals that the file has been touched we continue inside of the fork.

schneems commented:
I added a test, i'm pretty pleased with the result. Using pipes to synchronize the two processes.

schneems committed:
Lock the whole boot step, get rid of unneeded hash

schneems commented:
Updated to have a threadsafe `boot!` process, also using a instance variable instead of a `Concurrent::Hash`

sgrif commented:
Going to go forward with this one. I think my previous concerns were rooted in thinking that the `changed` callback did more than it actually does. I think we should eventually consider an explicit `Rails.application.forked!` call which needs to occur if you're using a forking web server, but that is out of scope for 5.0.0 at this point.

schneems commented:
Yay! Thanks! Just saw this notification in my email.

My only concern is that it's kinda weird that `update?` returns true on the first call. It makes sense for our application of the checker but it's an odd implementation artifact that maybe shouldn't live in this class.

As an alternative we could implement a `EventedFileChecker#forked?` that takes a block and use that in the middleware. Something like:

```ruby
evented_file_checker.forked? {|checker|
  checker.updated.make_true
  checker.boot!
end
```
